### PR TITLE
Gyro 2-CLP

### LINKED ISSUE #704: Port Gyro pools: 2-CLP
Initial effort in #420.

Pending: review why Liquidity Approximation tests are not passing.

### PR DESCRIPTION
**Pull request by joaobrunoah**
# Description

Gyro 2CLP pool, with all E2E and Liquidity Approximation tests in place.
## Type of change

- [ ] Bug fix 
- [x] New feature 
- [ ] Breaking change 
- [ ] Dependency changes
- [ ] Code refactor / cleanup
- [ ] Optimization: [ ] gas / [ ] bytecode
- [ ] Documentation or wording changes
- [ ] Other

## Checklist:

- [x] The diff is legible and has no extraneous changes
- [x] Complex code has been commented, including external interfaces
- [x] Tests have 100% code coverage
- [x] The base branch is either `main`, or there's a description of how to merge

## Issue resolution:

Closes #704

### CODE REVIEW

**Review by jubeira — COMMENTED** [review]
This looks pretty good; great job @joaobrunoah !

Just a couple of comments, but I see no strong reason not to approve this. I'll do one more pass before doing so anyways.

About the e2e tests: I had in mind that the change to compute invariant rounding down before and up after the operation was already in place, but that was not the case. I'll think about it a bit more but I think it's fine: probably 'standard' weighted pools don't need it because 50/50 is a special case where you don't really use `pow`, and therefore there's no real error. In general I believe this is the best we can do in solidity anyways.

Other than that this looks great. As a second step it'd be great to extend the fuzz tests as we did with weighted and stable to check different pool settings, but that's too much for the first iteration.

File: pkg/pool-gyro/contracts/lib/Gyro2CLPMath.sol [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+            uint256 bterm1 = _mulUpOrDown(balances[0], sqrtAlpha);
+            mb = bterm0 + bterm1;
+            // `c` is a term in the numerator and should be rounded up if we want to increase the invariant.
+            mc = _mulUpOrDown(balances[0], balances[1]);
+        }
+        // For better fixed point precision, calculate in expanded form re-ordering multiplications.
+        // `b^2 = x^2 * alpha + x*y*2*sqrt(alpha/beta) + y^2 / beta`
+        bSquare = _mulUpOrDown(_mulUpOrDown(balances[0], balances[0]), _mulUpOrDown(sqrtAlpha, sqrtAlpha));
+        uint256 bSq2 = _divUpOrDown(2 * _mulUpOrDown(_mulUpOrDown(balances[0], balances[1]), sqrtAlpha), sqrtBeta);
+        uint256 bSq3 = _divUpOrDown(_mulUpOrDown(balances[1], balances[1]), sqrtBeta.mulUp(sqrtBeta));
```
[the remark this answers is not in the document]
Comment (joaobrunoah) [reply pkg/pool-gyro/contracts/lib/Gyro2CLPMath.sol:94]: Agreed!

File: pkg/vault/test/foundry/E2eBatchSwap.t.sol [hunk 2]
```diff
@@ -137,31 +137,19 @@ contract E2eBatchSwapTest is BaseVaultTest {
         exactAmountIn = bound(exactAmountIn, minSwapAmountTokenA, maxSwapAmountTokenA);
 
         BaseVaultTest.Balances memory balancesBefore = getBalances(sender, tokensToTrack);
-        uint256[] memory invariantsBefore = _getPoolInvariants();
+        uint256[] memory invariantsBefore = _getPoolInvariants(Rounding.ROUND_DOWN);
```
[the remark this answers is not in the document]
Comment (joaobrunoah) [reply pkg/vault/test/foundry/E2eBatchSwap.t.sol:140]: Yes, it's in the E2eSwap.t.sol of Weighted Pool, but it wasn't implemented in the general tests.

**Review by jubeira — APPROVED** [review]
LGTM! I think this is a great starting point already.

The only missing thing would be to double down in liquidity approximation fuzz tests / e2e tests fuzzing pool-specific settings, but that's out of scope for this PR. Great job!

File: pkg/pool-gyro/contracts/lib/Gyro2CLPMath.sol [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+        }
+    }
+
+    /** @dev Calculate virtual offset a for reserves x, as in (x+a)*(y+b)=L^2
+     */
+    function _calculateVirtualParameter0(uint256 invariant, uint256 _sqrtBeta) internal pure returns (uint256) {
+        return invariant.divDown(_sqrtBeta);
+    }
+
+    /** @dev Calculate virtual offset b for reserves y, as in (x+a)*(y+b)=L^2
```
Comment (EndymionJkb) [inline pkg/pool-gyro/contracts/lib/Gyro2CLPMath.sol:231]:
```suggestion
    /** @dev Calculate the virtual offset `b` for reserves `y`, as in (x+a)*(y+b)=L^2
```

File: pkg/pool-gyro/contracts/lib/GyroPoolMath.sol [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    uint256 private constant _SQRT_1E_NEG_3 = 31622776601683793;
+    uint256 private constant _SQRT_1E_NEG_5 = 3162277660168379;
+    uint256 private constant _SQRT_1E_NEG_7 = 316227766016837;
+    uint256 private constant _SQRT_1E_NEG_9 = 31622776601683;
+    uint256 private constant _SQRT_1E_NEG_11 = 3162277660168;
+    uint256 private constant _SQRT_1E_NEG_13 = 316227766016;
+    uint256 private constant _SQRT_1E_NEG_15 = 31622776601;
+    uint256 private constant _SQRT_1E_NEG_17 = 3162277660;
+
+    /** @dev Implements square root algorithm using Newton's method and a first-guess optimisation **/
```
Comment (EndymionJkb) [inline pkg/pool-gyro/contracts/lib/GyroPoolMath.sol:22]:
```suggestion
    /** @dev Implements a square root algorithm using Newton's method and a first-guess optimization. **/
```

File: pkg/pool-gyro/contracts/lib/GyroPoolMath.sol [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    function sqrt(uint256 input, uint256 tolerance) internal pure returns (uint256) {
+        if (input == 0) {
+            return 0;
+        }
+
+        uint256 guess = _makeInitialGuess(input);
+
+        // At this point `guess` is an estimation with one bit of precision. We know the true value is a uint128,
+        // since it is the square root of a uint256. Newton's method converges quadratically (precision doubles at
+        // every iteration). We thus need at most 7 iteration to turn our partial result with one bit of precision
```
Comment (EndymionJkb) [inline pkg/pool-gyro/contracts/lib/GyroPoolMath.sol:32]:
```suggestion
        // every iteration). We thus need at most 7 iterations to turn our partial result with one bit of precision
```

File: pkg/pool-gyro/contracts/lib/GyroPoolMath.sol [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+        guess = (guess + ((input * FixedPoint.ONE) / guess)) / 2;
+        guess = (guess + ((input * FixedPoint.ONE) / guess)) / 2;
+        guess = (guess + ((input * FixedPoint.ONE) / guess)) / 2;
+        guess = (guess + ((input * FixedPoint.ONE) / guess)) / 2;
+        guess = (guess + ((input * FixedPoint.ONE) / guess)) / 2;
+
+        // Check squaredGuess (guess * guess) is close enough from input. `guess` has less than 1 wei error, but the
+        // loss of precision in the 18 decimal representation causes an error in the squared number, which must be
+        // less than `(guess * tolerance) / FixedPoint.ONE`. Tolerance, in this case, is a very small number (< 10), so
+        // the tolerance will be very small too.
```
Comment (EndymionJkb) [inline pkg/pool-gyro/contracts/lib/GyroPoolMath.sol:45]:
```suggestion
        // Check that squaredGuess (guess * guess) is close enough from input. `guess` has less than 1 wei error, but
        // the loss of precision in the 18 decimal representation causes an error in the squared number, which must be
        // less than `(guess * tolerance) / FixedPoint.ONE`. Tolerance, in this case, is a very small number (< 10),
        // so the tolerance will be very small too.
```

File: pkg/pool-gyro/coverage.sh [hunk 7]
```diff
@@ -0,0 +1,124 @@
+#!/bin/bash
```
Comment (EndymionJkb) [inline pkg/pool-gyro/coverage.sh:1]: Is this a soft-link? Or is it different and can't be?

File: pkg/pool-gyro/hardhat.config.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+import 'hardhat-gas-reporter';
+import 'hardhat-contract-sizer';
+import { warnings } from '@balancer-labs/v3-common/hardhat-base-config';
+
+const config: HardhatUserConfig = {
+  networks: {
+    hardhat: {
+      allowUnlimitedContractSize: true,
+    },
+  },
```
Comment (EndymionJkb) [inline pkg/pool-gyro/hardhat.config.ts:19]: Do we need unlimited contract size?

File: pkg/pool-gyro/test/foundry/utils/Gyro2ClpPoolDeployer.sol [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+contract Gyro2ClpPoolDeployer is BaseContractsDeployer {
+    using CastingHelpers for address[];
+
+    uint256 private _sqrtAlpha = 997496867163000167; // alpha (lower price rate) = 0.995
+    uint256 private _sqrtBeta = 1002496882788171068; // beta (upper price rate) = 1.005
+
+    string private artifactsRootDir = "artifacts/";
+
+    constructor() {
+        // if this external artifact path exists, it means we are running outside of this repo
```
Comment (EndymionJkb) [inline pkg/pool-gyro/test/foundry/utils/Gyro2ClpPoolDeployer.sol:29]:
```suggestion
        // If this external artifact path exists, it means we are running outside of this repo.
```

**Review by EndymionJkb — APPROVED** [review]
Just language/style things. Trusting others to have reviewed math, etc.

File: pkg/pool-gyro/coverage.sh (hunk 7 above)
Comment (joaobrunoah) [reply pkg/pool-gyro/coverage.sh:1]: https://stackoverflow.com/questions/8967902/why-do-you-need-to-put-bin-bash-at-the-beginning-of-a-script-file#:~:text=Adding%20%23!%2Fbin%2Fbash%20as,or%20%22sha%2Dbang%22.

This is just a convention, to tell the OS to run the script in bash.

**Pull request merged**